storage/innobase/handler/ha_innodb.h

```diff
@@ -492,7 +492,7 @@ class ha_innobase : public handler {
 
 #ifdef WITH_WSREP
   int wsrep_append_keys(THD *thd, Wsrep_service_key_type key_type,
```

Comment:
I am trying to understand whether wsrep_append_keys() should be in server or InnoDB.

it is required in InnoDB because of FKs, and also required in server because of triggers?

Reply:
Original idea is that writeset keys are added in two places:
1. During TOI start - there is a dedicated part
2. During DML execution - from InnoDB level. Every time the row is inserted/updated/deleted, from InnoDB level we add a key identifying the row.

The above idea is not enough, in case when on server level we detect that row will not be updated. In such a case we don't get to InnoDB world (we skip it), so certification key for the row that was considered to update is missing.
But we still have to add this cert key to the writeset for the reason described in commit message.